pytorch_to_returnn/torch/nn/modules/linear.py

```diff
@@ -129,12 +129,54 @@ class Dot(Module):
   is_original_torch_module = False
 
   def create_returnn_layer_dict(self, a: Tensor, b: Tensor, *, reduce_dim_a: int, reduce_dim_b: int) -> Dict[str, Any]:
-    assert a.shape == b.shape
+    from .operator import _unify_tensor_axes_returnn_meta
+    (a, b), _ = _unify_tensor_axes_returnn_meta(a, b)
```

Comment:
We should define what `Dot` is actually supposed to do.

>   Maps to RETURNN DotLayer for a special case

What does this really mean? What special case? Why only for a special case?

Using `_unify_tensor_axes_returnn_meta` only makes sense when `a` and `b` are supposed to have exactly the same axes, or maybe broadcast dims.

For `Dot`, this should not be needed. `DotLayer` in RETURNN categorizes the dims of `a` and `b` into three types:

- Axes to be reduced (need to be same in `a` and `b`) (broadcasting does not make sense)
- Shared axes, usually the batch axis
- Unique axes

So, e.g. if `a` is [B,T,F] and `b` is [B,F,G], when you reduce axis F, then B is shared and T and G are unique, and you get [B,T,G].

Given the reduced axis, we can automatically infer the shared axes (by just looking which dims occur in both `a` and `b`) and then the remaining axes are unique.

Maybe the shared axes are actually not supposed to be shared. E.g. [B,T,F] * [B,F,T], reduce=F, but you only want to share B, and then get [B,T,T]. To make that explicit, you can also specify `var1=T` and `var2=T`.

In any case, at no point it makes sense that `a` and `b` are exactly the same (`_unify_tensor_axes_returnn_meta`), and also broadcast dims don't make sense.

Reply:
> We should define what Dot is actually supposed to do.

Right, we should make that clearer. I added it because I needed it for the dot product in the cosine similarity. You suggested to allow making it more generic at some point. This is not necessary right now, but it probably helps to clarify the use of `Dot`.

Reply:
Yes, but again, as I said, there should never be the need for `_unify_tensor_axes_returnn_meta` or for broadcast dims.

Reply:
Right, I think I can just remove it.

Reply:
However, it's probably a good idea when we get explicit on `var1` and `var2`.

Reply:
You mean setting the non-matching axes explicitly as `var1` and `var2`?

Reply:
But what are non-matching axes? As I said, we first should define what this `Dot` layer is actually supposed to do (purely on PyTorch level, independent of RETURNN).

What are our current use cases?

Reply:
The only use case is in `torch.cosine_similarity` for the dot product along a specified axis. I'm fine with having this module just for this use case for now.

Reply:
I thought we had another use case? We introduced `Dot` already before that, or not?

What exactly is the use case in `torch.cosine_similarity`? What are the inputs `a` and `b`?

Reply:
In `torch.cosine_similarity`, we use it to compute the dot product of `a` and `b` along a given axis. `a` and `b` are two tensors which must be broadcastable to a common shape, see https://pytorch.org/docs/stable/generated/torch.nn.functional.cosine_similarity.html

Reply:
So this means, what we need for `Dot` is:

- Reduce axis is clear.
- Remove any broadcast dims before. I.e. iterate through the a/b shape and see if one of them is broadcast. If one of them is broadcast, add them to the corresponding `var1/2` of the other.
- The remaining axes should all be shared. But this is implicit when specifying `var1/2`.

The behavior is exactly equivalent to:
```
(a * b).sum(dim=reduce_dim)
```

This should be documented.

Reply:
Oh, right. We could just use `(a * b).sum(dim=reduce_dim)` in `torch.cosine_similarity` and remove `Dot` completely. I think this would be the cleanest thing as it's not used anywhere else.

Reply:
Ah I missed this message. But `(a * b).sum(dim=reduce_dim)` is less efficient, and requires more memory. Esp when the dim to be reduced is large. Maybe not so important here?